book/src/developers/designs/unixd_multi_resolver_2024.md

```diff
@@ -0,0 +1,306 @@
+│                                   ▲                 ▲
+                                    │                 │                         │
+│                                   ▼                 ▼
+                          ┌──────────────────┐   ┌────────┐                     │
+│                         │                  │   │        │
+                          │    DB / Cache    │   │  TPM   │                     │
+│                         │                  │   │        │
+                          └──────────────────┘   └────────┘                     │
+│
+ ─ ─ ─ ─ ─ ─ ─ ─ ─ ─ ── ─ ─ ─ ─ ─ ─ ─ ─ ─ ─ ─ ─ ─ ─ ─ ─ ─ ─ ─ ─ ─ ─ ─ ─ ─ ─ ─ ─ ┴
+```
+
+The current design treated the client as a trivial communication layer. The daemon/event loop
+contained all state including if the resolver was online or offline. Additionally the TPM and
+password caching operations primarily occured in the daemon layer, which limited the access of these
+features to the client backend itself.
+
+### Future Features
+
+#### Files Resolver
+
+The ability to resolve and authenticate local users from `/etc/{passwd,group,shadow}`. The classic
+mechanisms to resolve this are considered "slow" since they require a full-file-parse each
+operation.
+
+In addition, these files are limited by their formats and can not be extended with future
+authentication mechanisms like CTAP2 or PIV.
+
+Unixd already needs to parse these files to understand and prevent conflicts between local items and
+remote ones. Extending this functionality will allow us to resolve local users from memory.
+
+#### Local Group Extension
+
+An often requested feature is the ability to extend a local group with the members from a remote
+group. Often people attempt to achieve this by "overloading" a group remotely such as creating a
+group called "wheel" in Kanidm and then attempting to resolve it on their systems. This can
+introduce issues as different distributions may have the same groups but with different gidNumbers
+which can break systems, or it can cause locally configured items to be masked.
+
+Instead, we should allow group _extension_. A local group can be nominated for extension, and paired
+to a remote group. For example this could be configured as:
+
+```
+[group."wheel"]
+extend_from = "opensuse_wheel"
+```
+
+This allows the local group "wheel" to be resolved and _extended_ with the members from the remote
+group `opensuse_wheel`.
+
+#### Multiple Resolvers
+
+We would like to support multiple backends simultaneously and in our source tree. This is a major
+motivator of this rework as the himmelblau project wishes to contribute their client layer into our
+source tree, while maintaining the bulk of their authentication code in a separate libhimmelblau
+project.
+
+We also want to support LDAP and other resolvers too.
+
+The major challenge here is that this shift the cache state from the daemon to the client. This
```

Comment:
Since we (currently) expect to source-control all the clients, why not cache in the daemon and tag them with a client source?

Reply:
That's exactly what I'm proposing :)